### PR TITLE
Make move choice parsing flexible regarding targeting/flag order

### PR DESCRIPTION
**Pull request by scheibo**

### CODE REVIEW

**Comment by Zarel** [comment]
"Fix" is an interesting way to describe changing the order of the expected elements.

**Comment by Zarel** [comment]
It might be ideal to accept targeting/mega in any order, though.

**Comment by Zarel** [comment]
Also we should probably make choice failures throw in a test context.

**Comment by scheibo** [comment]
> "Fix" is an interesting way to describe changing the order of the expected elements.

`SIM-PROTOCOL.md` seems to imply `mega` comes last?

> `move MOVESPEC mega`, to mega-evolve and make a move
> `MOVESPEC` is:
> `MOVESLOTSPEC` or `MOVESLOTSPEC TARGETSPEC`

**Comment by Zarel** [comment]
It's way more likely that the protocol doc is wrong, than that the code is wrong.

**Comment by Zarel** [comment]
I think the ideal solution here is probably to accept targeting/mega in any order. Then we don't have to fix the doc _or_ the client!

**Comment by scheibo** [comment]
> I think the ideal solution here is probably to accept targeting/mega in any order. Then we don't have to fix the doc or the client!

OK, ill look into this. 

> Also we should probably make choice failures throw in a test context.

And this

**Comment by scheibo** [comment]
Alright, I've made it so its now agnostic to ordering, but `throw`ing breaks a lot of tests and is kind of unrelated to this PR so I've moved those changes to #5286 (my approach there might also be wrong, but we can talk about that over there).

**Comment by Zarel** [comment]
This pull request **introduces 1 alert** when merging b813fa6e220169c83c840effcd4fd431fe674c5c into 9e7ba68fc4caedf503638bd7edda6b781df18d65 - [view on LGTM.com](https://lgtm.com/projects/g/Zarel/Pokemon-Showdown/rev/pr-c1f391939b8a933ed0608f0ab4a1b1b3da7983f3)

**new alerts:**

* 1 for Useless assignment to local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: sim/side.ts [hunk 1]
```diff
@@ -629,17 +629,37 @@ export class Side {
...
+				// tslint:disable-next-line:one-variable-per-declaration
+				let targetLoc, willMega, willUltra, willZ;
+				// We're flexible with the ordering of target vs. flags - if `data `ends
+				// with a number then we assume target comes last, otherwise flags.
+				if (/[1-3]$/.test(data)) {
+					[targetLoc, data] = parseLoc(data);
+					[willMega, willUltra, willZ, data] = parseFlags(data);
+				} else {
+					[willMega, willUltra, willZ, data] = parseFlags(data);
+					[targetLoc, data] = parseLoc(data);
```
Comment (Zarel) [inline sim/side.ts]: This seems kind of inefficient.

```js
let targetLoc = 0;
let willMegaEtc = false;
while (true) {
	if (/\s-?[1-3]$/.test(data)) {
		targetLoc = parseInt(data.slice(-2), 10);
		data = data.slice(0, data.lastIndexOf(' '));
	} else if (data.endsWith(' mega')) {
		willMegaEtc = true;
		data = data.slice(0, -5);
	} else if (data.endsWith(' ultra')) {
		willMegaEtc = true;
		data = data.slice(0, -6);
	} else if (data.endsWith(' zmove')) {
		willMegaEtc = true;
		data = data.slice(0, -6);
	} else {
		break;
	}
}
```

Comment (scheibo) [reply sim/side.ts]: Done.

File: sim/side.ts [hunk 2]
```diff
@@ -629,17 +629,24 @@ export class Side {
 
 			switch (choiceType) {
 			case 'move':
-				let targetLoc = 0;
-				if (/\s-?[1-3]$/.test(data)) {
-					targetLoc = parseInt(data.slice(-2), 10);
-					data = data.slice(0, data.lastIndexOf(' '));
+				let [targetLoc, willMega, willUltra, willZ] = [0, '', '', ''];
```
Comment (Zarel) [inline sim/side.ts]: I still don't like this initialization.

```suggestion
				let targetLoc = 0;
				let megaOrZ = '';
```

File: sim/side.ts [hunk 3]
```diff
@@ -629,17 +629,24 @@ export class Side {
...
+				while (true) {
+					if (/\s-?[1-3]$/.test(data)) {
+						targetLoc = parseInt(data.slice(-2), 10);
+						data = data.slice(0, data.lastIndexOf(' '));
```
Comment (Zarel) [inline sim/side.ts]:
```suggestion
						data = data.slice(0, -2).trim();
```

In hindsight, this is probably easier to read.

File: sim/side.ts [hunk 4]
```diff
@@ -629,17 +629,24 @@ export class Side {
...
+						break;
+					}
 				}
-				const willMega = data.endsWith(' mega') ? 'mega' : '';
-				if (willMega) data = data.slice(0, -5);
-				const willUltra = data.endsWith(' ultra') ? 'ultra' : '';
-				if (willUltra) data = data.slice(0, -6);
-				const willZ = data.endsWith(' zmove') ? 'zmove' : '';
-				if (willZ) data = data.slice(0, -6);
 				this.chooseMove(data, targetLoc, willMega || willUltra || willZ);
```
Comment (Zarel) [inline sim/side.ts]:
```suggestion
				this.chooseMove(data, targetLoc, megaOrZ);
```

File: sim/side.ts (hunk 2 above)
Comment (Zarel) [reply sim/side.ts]: I actually explicitly think `willMega, willUltra, willZ` should be combined into one variable – it's done when `chooseMove` is called, anyway.

**Comment by Slayer95** [comment]
This doesn't reject invalid commands such as ``/move 1 1 mega ultra``. Unintended, right?

**Comment by Zarel** [comment]
The old code didn't, either. We could certainly throw if `megaOrZ` was truthy when we went to set it. Feel free to pullreq that change.

File: test/simulator/misc/choice-parser.js [hunk 5]
```diff
@@ -80,6 +80,21 @@ describe('Choice parser', function () {
...
+				]);
+				battle.join('p2', 'Guest 2', 1, [
+					{species: "Blaziken", ability: 'speedboost', item: 'blazikenite', moves: ['blazekick']},
+					{species: "Aggron", ability: 'sturdy', moves: ['irondefense']},
+				]);
+
+				assert.ok(battle.choose('p1', `move 1 1 mega, move smog 1`));
+				assert.ok(battle.choose('p2', `move 1 mega 2, move irondefense`));
+			});
+
```
Comment (Slayer95) [inline test/simulator/misc/choice-parser.js]: This test was placed under ``Switch requests > Doubles/Triples``. It should be under ``Move requests > Generic``

**Comment by Slayer95** [comment]
Oh you are right. The loop just made it more obvious.

**Comment by Zarel** [comment]
Where by "throw" I mean emit a choice error.

**Comment by scheibo** [comment]
> This test was placed under ``Switch requests > Doubles/Triples``. It should be under ``Move requests > Generic``

_Originally posted by @Slayer95 in https://github.com/Zarel/Pokemon-Showdown/pull/5284_


> Where by "throw" I mean emit a choice error.

_Originally posted by @Zarel in https://github.com/Zarel/Pokemon-Showdown/pull/5284#issuecomment-471244025_

Will follow up, gimme 5

**Pull request merged**